### PR TITLE
Support Cast JSON to REAL

### PR DESCRIPTION
**Pull request by ebyhr**
Added CAST JSON to MAP<REAL>
#7705

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java [hunk 1]
```diff
@@ -287,6 +287,9 @@ public void testJsonToMap()
         assertFunction("CAST(JSON '{\"1\":2.0, \"3\": 4.0}' AS MAP<BIGINT, DOUBLE>)",
                 new MapType(BIGINT, DOUBLE),
                 ImmutableMap.of(1L, 2.0, 3L, 4.0));
+        assertFunction("CAST(JSON '{\"1\":2.0, \"3\": 4.0}' AS MAP<BIGINT, REAL>)",
+                new MapType(BIGINT, REAL),
+                ImmutableMap.of(1L, (float) 2.8E-45, 3L, (float) 5.6E-45));
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java]: The test is wrong.

Also, can you add the test for CAST JSON TO REAL as well?

Comment (ebyhr) [inline presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java]: Sure, I will update it.

**Comment by ebyhr** [comment]
Sorry for the late reply. Updated.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 2]
```diff
@@ -224,6 +227,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+                    result = null;
+                    break;
+                case VALUE_STRING:
+                    result = VarcharOperators.castToFloat(Slices.utf8Slice(parser.getText()));
+                    break;
+                case VALUE_NUMBER_FLOAT:
+                    result = (long) floatToIntBits(parser.getFloatValue());
+                    break;
+                case VALUE_NUMBER_INT:
+                    result = (long) floatToIntBits(parser.getIntValue());
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: Why not using `getFloatValue()` here? (as we did for `castToDouble`? )

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 3]
```diff
@@ -224,6 +227,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+                    break;
+                case VALUE_TRUE:
+                    result = (long) floatToIntBits(BooleanOperators.castToFloat(true));
+                    break;
+                case VALUE_FALSE:
+                    result = (long) floatToIntBits(BooleanOperators.castToFloat(false));
+                    break;
+                default:
+                    throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), DOUBLE));
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: Change the last parameter from `DOUBLE` to `REAL`. Do the same for line 262

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 4]
```diff
@@ -331,6 +372,23 @@ public static Slice castFromDouble(@SqlType(DOUBLE) double value)
 
     @ScalarOperator(CAST)
     @SqlType(JSON)
+    public static Slice castFromReal(@SqlType(REAL) long value)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: Let's do this (CAST REAL TO JSON) in a separate PR

File: presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java [hunk 5]
```diff
@@ -100,6 +100,13 @@ public static double castToDouble(@SqlType(StandardTypes.BOOLEAN) boolean value)
     }
 
     @ScalarOperator(CAST)
+    @SqlType(StandardTypes.REAL)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: Let's not do it in this PR.

File: presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java [hunk 6]
```diff
@@ -287,8 +294,8 @@ else if (javaType == long.class) {
             if (element instanceof SqlDecimal) {
                 type.writeLong(blockBuilder, ((SqlDecimal) element).getUnscaledValue().longValue());
             }
-            else if (REAL.equals(type)) {
-                type.writeLong(blockBuilder, floatToRawIntBits(((Number) element).floatValue()));
+            else if (type.equals(REAL)) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java]: why change to `floatToIntBits?`

File: presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java [hunk 7]
```diff
@@ -149,7 +149,12 @@ else if (type.getJavaType() == boolean.class) {
             type.writeBoolean(blockBuilder, parser.getBooleanValue());
         }
         else if (type.getJavaType() == long.class) {
-            type.writeLong(blockBuilder, parser.getLongValue());
+            if (type.equals(REAL)) {
+                type.writeLong(blockBuilder, floatToIntBits(parser.getFloatValue()));
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java]: use `floatToRawIntBits`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 8]
```diff
@@ -224,6 +227,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+            Long result;
+            switch (parser.getCurrentToken()) {
+                case VALUE_NULL:
+                    result = null;
+                    break;
+                case VALUE_STRING:
+                    result = VarcharOperators.castToFloat(Slices.utf8Slice(parser.getText()));
+                    break;
+                case VALUE_NUMBER_FLOAT:
+                    result = (long) floatToIntBits(parser.getFloatValue());
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: use `floatToRawIntBits`

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 9]
```diff
@@ -147,6 +148,10 @@ public void testArrayToJson()
         assertFunction("cast(cast(ARRAY[123456789, -123456789, null] AS ARRAY<INTEGER>) AS JSON)", JSON, "[123456789,-123456789,null]");
         assertFunction("cast(cast(ARRAY[1234567890123456789, -1234567890123456789, null] AS ARRAY<BIGINT>) AS JSON)", JSON, "[1234567890123456789,-1234567890123456789,null]");
 
+        assertFunction("cast(cast(ARRAY[1, 2, null] AS ARRAY<REAL>) AS JSON)", JSON, "[1.0,2.0,null]");
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: These tests are not necessary in this PR. `ARRAY<REAL> TO JSON` is already supported.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 10]
```diff
@@ -189,6 +194,8 @@ public void testJsonToArray()
         assertFunction("CAST(JSON '[1, null, 3]' AS ARRAY<BIGINT>)", new ArrayType(BIGINT), asList(1L, null, 3L));
         assertFunction("CAST(JSON '[1, 2.0, 3]' AS ARRAY<DOUBLE>)", new ArrayType(DOUBLE), ImmutableList.of(1.0, 2.0, 3.0));
         assertFunction("CAST(JSON '[1.0, 2.5, 3.0]' AS ARRAY<DOUBLE>)", new ArrayType(DOUBLE), ImmutableList.of(1.0, 2.5, 3.0));
+        assertFunction("CAST(JSON '[1, 2, 3]' AS ARRAY<REAL>)", new ArrayType(REAL), ImmutableList.of(1.0f, 2.0f, 3.0f));
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Can we have decimals here? (e.g. 2.5 )

File: presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java [hunk 11]
```diff
@@ -115,6 +116,43 @@ public void testCastFromDouble()
...
+        assertInvalidFunction("cast(JSON '\"true\"' as REAL)", INVALID_CAST_ARGUMENT);
+
+        assertFunction("cast(JSON ' 128.9' as REAL)", REAL, 128.9f); // leading space
+
+        assertFunction("cast(json_extract('{\"x\":1.23}', '$.x') as REAL)", REAL, 1.23f);
+        assertInvalidCast("cast(JSON '{ \"x\" : 123}' as REAL)");
+    }
+
+    @Test
+    public void testCastFromReal()
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java]: Not for this PR :)

File: presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java (hunk 6 above)
Comment (ebyhr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java]: `floatToRawIntBits ` fails in following test. I'm investigating the cause.

```java
assertFunction("cast(JSON 'true' as REAL)", REAL, 1.0f);
```

**Comment by ebyhr** [comment]
Updated.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 12]
```diff
@@ -224,6 +226,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+                    result = VarcharOperators.castToFloat(Slices.utf8Slice(parser.getText()));
+                    break;
+                case VALUE_NUMBER_FLOAT:
+                    result = (long) floatToRawIntBits(parser.getFloatValue());
+                    break;
+                case VALUE_NUMBER_INT:
+                    result = (long) floatToRawIntBits(parser.getFloatValue());
+                    break;
+                case VALUE_TRUE:
+                    result = (long) floatToRawIntBits(1);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: We should add `CAST BOOLEAN TO REAL` support and use `BooleanOperators.castToReal(true)`. Do this in a separate commit.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 13]
```diff
@@ -224,6 +226,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+                    result = (long) floatToRawIntBits(0);
+                    break;
+                default:
+                    throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), REAL));
+            }
+            checkCondition(parser.nextToken() == null, INVALID_CAST_ARGUMENT, "Cannot cast input json to REAL"); // check no trailing token
+            return result;
+        }
+        catch (IOException e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), DOUBLE));
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: REAL

**Comment by wenleix** [comment]
@ebyhr: Also, once you finish the round, reassign it back to me since otherwise it's not easy for me to find outstanding PR waiting for request on GitHub :(

**Comment by cawallin** [comment]
I'm pretty sure you need `floatToRawIntBits` -- @maciejgrzybek can confirm

**Comment by maciejgrzybek** [comment]
We use `floatToRawIntBits` throughout Presto.
The difference is in representation of NaN values. `floatToRawIntBits` preserves actual bit representation of NaN, while `floatToIntBits` returns the same value for all NaNs - `0x7fc00000`.

**Comment by ebyhr** [comment]
@maciejgrzybek @cawallin  Thanks for the comments. I use `floatToRawIntBits`.

@wenleix I added boolean to real cast. Probably, I don't have the permission to change the assign in this repository.

**Comment by ebyhr** [comment]
FYI: The build(`ea92fb8`) successes in my personal TravisCI.
https://travis-ci.org/ebyhr/presto/builds/224911521

**Comment by wenleix** [comment]
@ebyhr Oh I see :) I restarted the Jenkins :)

File: presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java [hunk 14]
```diff
@@ -100,6 +100,13 @@ public static double castToDouble(@SqlType(StandardTypes.BOOLEAN) boolean value)
...
+    public static long castToReal(@SqlType(StandardTypes.BOOLEAN) boolean value)
+    {
+        return value ? 1065353216 : 0; // 1065353216 means 1.0f
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: Sorry now I see why you are adding CAST BOOLEAN to REAL in the previous commits. You can directly discuss with me if you feel I misunderstand anything :). 

Let's do this:
1. Make this commit before the CAST JSON TO REAL COMMIT
2. Add separate test cases for CAST BOOLEAN to REAL

Sorry about the confusion!

For the code, what about this? :

```
return value ? floatToRawIntBits(1.0) : floatToRawIntBits (0.0);
```

Comment (ebyhr) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: Oh Sorry, I misunderstood. I'm okay now :)

I'll modify the code.

**Comment by ebyhr** [comment]
@wenleix I'm not on the Presto Slack (Is it open channel?). Please feel free to add my account :)
Slack username / address: @ebyhr / ebyhry@gmail.com

File: presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java [hunk 15]
```diff
@@ -100,6 +101,13 @@ public static double castToDouble(@SqlType(StandardTypes.BOOLEAN) boolean value)
     }
 
     @ScalarOperator(CAST)
+    @SqlType(StandardTypes.REAL)
+    public static long castToReal(@SqlType(StandardTypes.BOOLEAN) boolean value)
+    {
+        return value ? floatToRawIntBits(1) : floatToRawIntBits(0);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: I personally prefer to use `1.0f` and `0.0f` here. @haozhun what's your opinion? (We are using `1` and `0` in `castToDouble`, though.

Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: I believe `1`/`0` is fine. But I see no downside in writing `1.f`/`1.0f` and `0.f`/`0.0f`. I don't really care.

File: presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java [hunk 16]
```diff
@@ -100,6 +101,13 @@ public static double castToDouble(@SqlType(StandardTypes.BOOLEAN) boolean value)
     }
 
     @ScalarOperator(CAST)
+    @SqlType(StandardTypes.REAL)
+    public static long castToReal(@SqlType(StandardTypes.BOOLEAN) boolean value)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/type/BooleanOperators.java]: Please change the commit message to `Add CAST from BOOLEAN to REAL`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 17]
```diff
@@ -224,6 +226,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
 
     @ScalarOperator(CAST)
     @SqlNullable
+    @SqlType(REAL)
+    public static Long castToFloat(@SqlType(JSON) Slice json)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: Please change the commit message to `Add CAST from JSON to REAL`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 18]
```diff
@@ -224,6 +226,44 @@ public static Double castToDouble(@SqlType(JSON) Slice json)
...
+                    result = null;
+                    break;
+                case VALUE_STRING:
+                    result = VarcharOperators.castToFloat(Slices.utf8Slice(parser.getText()));
+                    break;
+                case VALUE_NUMBER_FLOAT:
+                    result = (long) floatToRawIntBits(parser.getFloatValue());
+                    break;
+                case VALUE_NUMBER_INT:
+                    result = (long) floatToRawIntBits(parser.getFloatValue());
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: Add comment here for justification (just like we did in `castToDouble`).

Even though I wrote that code (in `castToDouble` originally), when I saw this line, it's not obvious at all to me why we are doing this (instead of `BigintOperators.castToFloat`).

File: presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java [hunk 19]
```diff
@@ -115,6 +116,34 @@ public void testCastFromDouble()
...
+            throws Exception
+    {
+        assertFunction("cast(JSON 'null' as REAL)", REAL, null);
+        assertFunction("cast(JSON '-128' as REAL)", REAL, -128.0f);
+        assertFunction("cast(JSON '128' as REAL)", REAL, 128.0f);
+        assertFunction("cast(JSON '1.234567890' as REAL)", REAL, 1.234567890f);
+        assertFunction("cast(JSON '128.9' as REAL)", REAL, 128.9f);
+        assertFunction("cast(JSON '1e-324' as REAL)", REAL, 0.0f); // smaller than minimum subnormal positive
+        assertFunction("cast(JSON '1e309' as REAL)", REAL, Float.POSITIVE_INFINITY); // overflow
+        assertFunction("cast(JSON '-1e309' as REAL)", REAL, Float.NEGATIVE_INFINITY); // underflow
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java]: 1e-324, 1e309 and -1e309 need to be adjusted for 32-bit float.

File: presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java [hunk 20]
```diff
@@ -115,6 +116,34 @@ public void testCastFromDouble()
...
+        assertFunction("cast(JSON 'true' as REAL)", REAL, 1.0f);
+        assertFunction("cast(JSON 'false' as REAL)", REAL, 0.0f);
+        assertFunction("cast(JSON '\"128\"' as REAL)", REAL, 128.0f);
+        assertFunction("cast(JSON '\"12345678901234567890\"' as REAL)", REAL, 1.2345678901234567e19f);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java]: `1.2345678901234567e19f` is well beyond the accuracy of float. Shorten it so that it makes more sense.

**Comment by ebyhr** [comment]
@haozhun Thanks for your review :)

@wenleix Updated.

File: presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java (hunk 20 above)
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java]: @haozhun The goal of this test is to test when the underlying integer is overflowing BIGINT, it should still successfully convert into a REAL number although with precision loss.

**Comment by wenleix** [comment]
Looks good. I will merge it with some small change after the release :)

File: presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java [hunk 21]
```diff
@@ -115,6 +116,34 @@ public void testCastFromDouble()
     }
 
     @Test
+    public void testCastToReal()
+            throws Exception
+    {
+        assertFunction("cast(JSON 'null' as REAL)", REAL, null);
+        assertFunction("cast(JSON '-128' as REAL)", REAL, -128.0f);
+        assertFunction("cast(JSON '128' as REAL)", REAL, 128.0f);
+        assertFunction("cast(JSON '1.234567890' as REAL)", REAL, 1.234567890f);
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/type/TestJsonOperators.java]: We mean this:
```
assertFunction("cast(JSON '12345678901234567890' as REAL)", REAL, 1.2345679e19f);
```

The goal to have this huge number is to make sure it doesn't first cast into an BIGINT (which will cause Overflow). 

I will change it for the merge :)

**Comment by wenleix** [comment]
Merged #7740. Thanks!
